### ISSUE TITLE
conda-build 2.x produces defunct builds for Python version constraints

### ISSUE THREAD

**Issue by mbargull** [conda/conda-build#1962]
\[Clarification: This issue is about recipes that **do not** use `noarch: python`.\]
\[Clarification edit/correction: The issue actually also affects `noarch: python` which include scripts in `bin`/`Scripts`.\]

Note `conda-build 3.0.0beta1` is not affected by this and works perfectly fine.
`conda-build 2.x` produces broken packages for the following:
```yaml
# conda.recipe/meta.yaml
package:
  name: pyver-constraint
  version: 0.0.1

source:
  path: .

build:
  script: python setup.py install --single-version-externally-managed --record=record.txt

requirements:
  build:
    - python
    - setuptools
  run:
    - python >=3.4  # here the ">="-constraint is causing the issue
```
```python
# setup.py
from setuptools import setup
setup(name='pyver-constraint', version='0.0.1', py_modules=['some_empty_module'])
```
```python
# some_empty_module.py
```
The version constraint `>=3.4` prevents conda-build from creating a build specific to the given Python version due to
```python
def build_string_from_metadata(metadata):
...
                if any(i in v for i in ',|>!<'):
                    break
...
```
in [metadata.py#L389-L390](https://github.com/conda/conda-build/blob/2.1.x/conda_build/metadata.py#L389-L390), I guess. The build process runs without complaints but creates an invalid output. See below.
```bash
$ cd conda.recipe
$ conda-build --no-anaconda-upload --python 3.5 .
BUILD START: pyver-constraint-0.0.1-0
updating index in: /home/maba/miniconda3/conda-bld/linux-64
updating index in: /home/maba/miniconda3/conda-bld/noarch

The following NEW packages will be INSTALLED:

    openssl:    1.0.2k-1     
    pip:        9.0.1-py35_1 
    python:     3.5.3-1      
    readline:   6.2-2        
    setuptools: 27.2.0-py35_0
    sqlite:     3.13.0-0     
    tk:         8.5.18-0     
    wheel:      0.29.0-py35_0
    xz:         5.2.2-1      
    zlib:       1.2.8-3      

Copying /home/maba/code/conda/conda-build/issues/pyver-constraint to /home/maba/miniconda3/conda-bld/pyver-constraint_1493243807227/work
Package: pyver-constraint-0.0.1-0
source tree in: /home/maba/miniconda3/conda-bld/pyver-constraint_1493243807227/work
+ source /home/maba/miniconda3/bin/activate /home/maba/miniconda3/conda-bld/pyver-constraint_1493243807227/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place
+ python setup.py install --single-version-externally-managed --record=record.txt
running install
running build
running build_py
creating build
creating build/lib
copying some_empty_module.py -> build/lib
running install_lib
copying build/lib/some_empty_module.py -> /home/maba/miniconda3/conda-bld/pyver-constraint_1493243807227/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.5/site-packages
byte-compiling /home/maba/miniconda3/conda-bld/pyver-constraint_1493243807227/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.5/site-packages/some_empty_module.py to some_empty_module.cpython-35.pyc
running install_egg_info
running egg_info
creating pyver_constraint.egg-info
writing dependency_links to pyver_constraint.egg-info/dependency_links.txt
writing top-level names to pyver_constraint.egg-info/top_level.txt
writing pyver_constraint.egg-info/PKG-INFO
writing manifest file 'pyver_constraint.egg-info/SOURCES.txt'
reading manifest file 'pyver_constraint.egg-info/SOURCES.txt'
writing manifest file 'pyver_constraint.egg-info/SOURCES.txt'
Copying pyver_constraint.egg-info to /home/maba/miniconda3/conda-bld/pyver-constraint_1493243807227/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.5/site-packages/pyver_constraint-0.0.1-py3.5.egg-info
running install_scripts
writing list of installed files to 'record.txt'
INFO conda_build.build:bundle_conda(865): Packaging pyver-constraint-0.0.1-0
compiling .pyc files...
number of files: 6
Fixing permissions
Fixing permissions
/home/maba/miniconda3/conda-bld/linux-64/pyver-constraint-0.0.1-0.tar.bz2
updating index in: /home/maba/miniconda3/conda-bld
updating index in: /home/maba/miniconda3/conda-bld/noarch
updating index in: /home/maba/miniconda3/conda-bld/linux-64
updating: pyver-constraint-0.0.1-0.tar.bz2
updating index in: /home/maba/miniconda3/conda-bld
Nothing to test for: /home/maba/miniconda3/conda-bld/linux-64/pyver-constraint-0.0.1-0.tar.bz2
INFO conda_build.config:__exit__(432): --dirty flag not specified.  Removing build folder after successful build/test.

# Automatic uploading is disabled
# If you want to upload package(s) to anaconda.org later, type:

anaconda upload /home/maba/miniconda3/conda-bld/linux-64/pyver-constraint-0.0.1-0.tar.bz2

# To have conda build upload to anaconda.org automatically, use
# $ conda config --set anaconda_upload yes

anaconda_upload is not set.  Not uploading wheels: []
$ tar -xOf ~/miniconda3/conda-bld/linux-64/pyver-constraint-0.0.1-0.tar.bz2 info/files
lib/python3.5/site-packages/pyver_constraint-0.0.1-py3.5.egg-info/dependency_links.txt
lib/python3.5/site-packages/__pycache__/some_empty_module.cpython-35.pyc
lib/python3.5/site-packages/pyver_constraint-0.0.1-py3.5.egg-info/PKG-INFO
lib/python3.5/site-packages/some_empty_module.py
lib/python3.5/site-packages/pyver_constraint-0.0.1-py3.5.egg-info/SOURCES.txt
lib/python3.5/site-packages/pyver_constraint-0.0.1-py3.5.egg-info/top_level.txt
$ tar -xOf ~/miniconda3/conda-bld/linux-64/pyver-constraint-0.0.1-0.tar.bz2 info/index.json
{
  "arch": "x86_64",
  "build": "0",
  "build_number": 0,
  "depends": [
    "python >=3.4"
  ],
  "name": "pyver-constraint",
  "platform": "linux",
  "subdir": "linux-64",
  "version": "0.0.1"
}
$ conda create --use-local -y -n pyver-constraint --python=3.6 pyver-constraint
...
$ source activate pyver-constraint
$ python -m some_empty_module
/home/maba/miniconda3/envs/pyver-constraint/bin/python: No module named some_empty_module
```
Depending on how long `2.x` is maintained, this might needn't be entirely fixed, but the build should at least fail. Sadly I don't know the code too well and thus won't be able to issue a PR.

**Comment by msarahan** [comment]
Thanks. This issue has been cropping up a lot lately. I don't think it is a regression, but it very well might be.

I'll add your recipe here as a test and see if I can fix 2.1.x. 

Thanks for testing 3.0, too!

**Comment by ma-ba** [comment]
Correct, this is not a regression. I had some older builds from 1.2.1 that I just recently noticed to be affected by this. Hadn't had the chance to test 3.0 apart from this, but am looking forward to it!

**Comment by msarahan** [comment]
cc @stuartarchibald - this is the same issue you reported.

**Comment by stuartarchibald** [comment]
@msarahan This is very similar. IIRC I had something a bit like:
```YAML
package:
  name: foo

source:
  path: ..

build:
  script: python setup.py install

requirements:
  build:
    - python ==3.5
    - setuptools
  run:
    - python >=3.5

test:
  requires:
    -pytest
  commands:
    - python -m pytest
```
This is clearly a mistake given the internal workings, but worked until py36 appeared at which point I think 3.5 was used at build time and then at test (run) time a 3.6 env was being used which actually meant the tests failed as there was no module to import for py36 as the package was built for py35.

This may be a different issue to that of the OP/there are other things about this which perhaps ought be tracked separately (can create more tickets if that's best).
 1. Pinned python version is  bad (perhaps emit a large warning from the parser or ban it?!).
 2. If fixes are put in place to handle pinning, the pin is respected in `build` but doesn't propagate.
 3. In general `>=` constraints do not work for python (perhaps raise if encountered?).

**Comment by ma-ba** [comment]
Right, this is the same issue, i.e., `run: [python >=3.5]` being the problem. I noticed the problem even later than the py35 -> py36 transition since my user base are mostly Windows users. This issue isn't necessarily noticeable on Windows since there are no `lib/python?.?/` sub directories unlike on the other platforms.
Regarding @stuartarchibald 's points:
_Constrained_ Python versions can be handy. I altered my packages to be built with
```yaml
{% set data = load_setup_py_data() %}
build:
  - noarch: python
requirements:
  run:
    - python {{ data['python_requires'] }}
```
since they run cross-platform but also use features from `python >=3.4`. This runs marvellous.
So if any of those restrictions are put in place, these should only concern platform dependent builds.
